Send a Single Email to a Single Recipient #331

M3LiNdRu opened the pull request:
This pull request doesn´t solve the issue #331. But it´s an inflexion point to start following the same arquitecture and conventions. Due to the refactoring I'm applying to the source code I prefered split the issue in small Pull Request.

In this PR I included:
- Added new project for Integration Test using AAA pattern
- Started refactor in SendGridAPIClient to follow SOLID principles
- Created Mail Service to implement operations related to Mail

M3LiNdRu commented:
@thinkingserious  I would like to discuss a couple of architecture concepts.

- Following the single responsibility principle. I think Client class should be responsible for handling all HTTP stuff and API Services (Mail Service) shouldn't deal with any HTTP concept like Request or Response. EmailService only would send his Custom Request object to Client and then Client would has a internal serializer to serialize/deserialize Custom Request/Response objects.

- Also, In my opinion, the mail helper has too much responsabilities. We should be refactored.

If you are agree with the above points, I don't have any problem to create issues for covering that.

Many Thanks

thinkingserious commented:
Hello @M3LiNdRu,

I agree with your comments on the responsibility of the Client class.

What responsibilities do you think the mail helper should cover?

M3LiNdRu commented:
In my opinion, regarding the api v3 reference, the mail helper it would be responsible of the [mail request body parameters.](https://sendgrid.com/docs/API_Reference/Web_API_v3/Mail/index.html#-Request-Body-Parameters) I think the mail helper should be refactored as an object, not as a helper. It contains all mail info. 

@thinkingserious  What do you think?

M3LiNdRu commented:
That PR fix the issue #331. Changes applied:

1. Refactored Client.cs, it's splitted in two classes (SendGridAPI, SendGridHTTPClient). SendGridAPI contains the API and SendGridHttpClient all HTTP stuff. A new method has been created for posting data (PostAsync) to send grid api.

2. Refactor Mail Helper. It´s splitted in two classes as well (Models.Mail, MailExtensions) Models.Mail contains the real object described in the SendGrid API Reference. MailExtensions contains all methods related to the Models.Mail

3. Changed some names and following a convention. Right now, extend the code with more funcionalities is really handy and it doesn't require too much overwork. Just creating the interface/service, isolate the object to Models folder and registering the service in SendGridAPI class

thinkingserious commented on SendGrid/Example/Example.cs:
```diff
@@ -35,13 +35,13 @@ private static void Main()
             Email to = new Email("elmer@sendgrid.com");
             Content content = new Content("text/html", "I'm replacing the <strong>body tag</strong>");
-            Mail mail = new Mail(from, subject, to, content);
+            SendGrid.Models.Mail mail = new SendGrid.Models.Mail(from, subject, to, content);
```
I'm thinking we move this code behind the scenes.

I think something like this would replace this example:

```csharp
String apiKey = Environment.GetEnvironmentVariable("SENDGRID_APIKEY", EnvironmentVariableTarget.User);
email = new Email(SendGrid.SendGridAPIClient(apiKey));
// The internal Email object would be renamed to Contact and the object would be instantiated behind the scenes, same with all the other getters and setters.
email.SetFrom("dx@sendgrid.com");
email.AddSubject("I'm replacing the subject tag");
email.AddContent("text/html", "I'm replacing the <strong>body tag</strong>");
email.SetTemplateID("13b8f94f-bcae-4ec6-b752-70d6cb59f932");
email.AddSubstitution("-name-", "Example User");
email.AddSubstitution("-city-", "Denver");
if(email.Verify() == True){
  response = email.Send();
}
Console.WriteLine(response);
Console.ReadLine();
```

M3LiNdRu replied:
I agree to inject the Api client as a dependency, but I think it should be injected in the EmailService not in the Email object. In my opinion, Email object should be just a DTO.

thinkingserious commented:
Please see my comments in the review, and following are my thoughts on your previous comment:

_Refactored Client.cs, it's splitted in two classes (SendGridAPI, SendGridHTTPClient). SendGridAPI contains the API and SendGridHttpClient all HTTP stuff. A new method has been created for posting data (PostAsync) to send grid api._

Please see my comments along with the branch I just merged into v9beta to checkout the direction I'm thinking. 

_Refactor Mail Helper. It´s splitted in two classes as well (Models.Mail, MailExtensions) Models.Mail contains the real object described in the SendGrid API Reference. MailExtensions contains all methods related to the Models.Mail_

I like the direction you are going. I'll dig in deeper when I finish my current tasks.

_Changed some names and following a convention. Right now, extend the code with more funcionalities is really handy and it doesn't require too much overwork. Just creating the interface/service, isolate the object to Models folder and registering the service in SendGridAPI class._

Thanks for all of your help thus far. I'll be able to collaborate with you closer once I finish up my current tasks. That said, please do continue the conversation.

thinkingserious commented:
Here is my working pull request/branch: https://github.com/sendgrid/sendgrid-csharp/pull/356

M3LiNdRu commented on SendGrid/Example/Example.cs:
```diff
@@ -35,13 +35,13 @@ private static void Main()
             mail.Personalization[0].AddSubstitution("-city-", "Denver");
 
-            dynamic response = await sg.client.mail.send.post(requestBody: mail.Get());
+            dynamic response = await sg.client.mail.send.post(requestBody: mail.Serialize());
```
@thinkingserious I'd say I'm not able to get this change until you accept your pull request and you merge it to the branch

thinkingserious replied:
Correct, that is coming soon. I apologize for the delay. I had a couple of days of PTO and I'm currently trying to catch up. Please stay tuned ...

jduncanator commented:
Looking at this very briefly, it doesn't look like it will work very well (if at all) in multi-threaded applications. Why do you assign the mail request to the client and then call `SendAsync`? Why not pass the Mail model into the `SendAsync` method and send it from there. At least the current implementation can share a Sendgrid Client across threads. With the code in it's current state, I have to implement locking on the MailRequest member so as to make sure I do not overwrite a MailRequest from one thread whilst another thread is currently sending an email.

I personally think this is a step in the wrong direction.

jduncanator commented on SendGrid/SendGrid.IntegrationTests/MailServiceTest.cs:
```diff
@@ -0,0 +1,64 @@
+
+
+            //Act
+            _client.Mail.MailRequest = mail;
```
How will this work in multi-threaded applications? If one thread writes to this variable whilst another thread is sending an email, you have a recipe for disaster. In it's current state it would be up to the consumer to implement locking on this variable. This should be handled internally and the library should be completely thread-safe.

chadwackerman commented:
I'm sorry if any of this comes of as inappropriate but I cannot believe the amount of effort that has been expended in this thread and the end result as @jduncanator points out is junior engineering bugs.

Can you please just hire a single experienced C# dev to write this API for you? This thing is totally off in the weeds. It's taken a month for someone sane to step in and say "uh, this isn't how anyone else does it guys." Good lord.

thinkingserious commented:
@chadwackerman @jduncanator,

I'll be coming back to this thread shortly, I need to finish this first: https://github.com/sendgrid/sendgrid-csharp/pull/356
